### PR TITLE
Send envelopes that cannot be cached to disk

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

Try to send envelopes that could not be saved to the disk cache.

## :bulb: Motivation and Context

Relates to https://github.com/getsentry/team-mobile/issues/52

## :green_heart: How did you test it?

- Added unit test

## :pencil: Checklist

You have to check all boxes before merging:

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

## :crystal_ball: Next steps

Check thread syncing.

### CODE REVIEW

File: Sources/Sentry/SentryHttpTransport.m
Comment (philipphofmann) [inline Sources/Sentry/SentryHttpTransport.m:142]: `h`: We will run into threading issues here. `sendAllCachedEnvelopes` can be called from multiple threads, but it's thread-safe by ensuring it runs only once. Still, it can happen that `sendAllCachedEnvelopes` removes an object from `notStoredEnvelopes` on a different thread while we add an object here. To solve this, I suggest adding `@synchronized` when accessing `notStoredEnvelopes`. The extra lock is acceptable as it's an edge case and it's executed on a bg thread.

File: Sources/Sentry/SentryHttpTransport.m
Comment (philipphofmann) [inline Sources/Sentry/SentryHttpTransport.m:140]: `m`: `storeEnvelope` already has the info when it fails to store the envelope

https://github.com/getsentry/sentry-cocoa/blob/24e07440d85b8d07ae4a98885d0f8f440ede339e/Sources/Sentry/SentryFileManager.m#L313-L328

 We could avoid the extra call to `fileExistsAtPath` by either returning `nil` or changing the methods return value to a boolean. The SDK doesn't use the path returned from the method except here.

File: Sources/Sentry/SentryHttpTransport.m
Comment (philipphofmann) [inline Sources/Sentry/SentryHttpTransport.m:340]: `l`:
```suggestion
- (void)deleteEnvelopeAndSendNext:(nullable NSString *)envelopePath
```

**Review by philipphofmann — COMMENTED** [review]
Thanks for doing this. A few suggestions.

File: Sources/Sentry/SentryHttpTransport.m
Comment (denrase) [reply Sources/Sentry/SentryHttpTransport.m]: We use the path in tests. I've opted to return nil if we cannot save. That way the old functionality is preserved and we can still check whether the method failed from the call side.

File: Sources/Sentry/SentryFileManager.m
Comment (philipphofmann) [inline Sources/Sentry/SentryFileManager.m:324]: `m`: We can `return nil` here to avoid calling `handleEnvelopesLimit`.

File: Sources/Sentry/SentryHttpTransport.m
Comment (philipphofmann) [inline Sources/Sentry/SentryHttpTransport.m:299]: `m`: I think we should only limit the synchronized block to this code. Especially `deleteEnvelopeAndSendNext` shouldn't be in the synchronized block. Something like this could do the trick:

```
@synchronized(self.notStoredEnvelopes) {
    if (self.notStoredEnvelopes.count > 0) {
        envelope = self.notStoredEnvelopes[0];
        [self.notStoredEnvelopes removeObjectAtIndex:0];
    }
}

if (envelope == nil) {
...
```

File: Tests/SentryTests/Networking/SentryHttpTransportTests.swift
Comment (philipphofmann) [inline Tests/SentryTests/Networking/SentryHttpTransportTests.swift:490]: `m`: With https://github.com/getsentry/sentry-cocoa/pull/4337 you can now create an envelope that will fail serialization / can't be stored to disk; see https://github.com/getsentry/sentry-cocoa/blob/888a145b144b8077e03151a886520f332e47e297/Tests/SentryTests/Helper/SentryFileManagerTests.swift#L137-L144

So then you don't need the TestFileManager and can use the normal FileManager, which is a more realistic use case.

**Review by philipphofmann — COMMENTED** [review]
Almost there 🥅

File: Tests/SentryTests/Networking/SentryHttpTransportTests.swift
Comment (denrase) [reply Tests/SentryTests/Networking/SentryHttpTransportTests.swift:490]: We already use the `TestFileManger` in the `SentryHttpTransportTests`. IMHO it makes sense here to test against the interface, not an implementations. From the transports test it does not matter why the method retuned nil, just that it did.

**Review by philipphofmann — COMMENTED** [review]
Sorry that I didn't see this earlier. The SentryHttpTransport calls https://github.com/getsentry/sentry-cocoa/blob/394ad331f5d7c010ae68d884feda0a71dff5b194/Sources/Sentry/SentryNSURLRequestBuilder.m#L15-L18

The `SentrySerialization dataWithEnvelope` can return nil, but the SentryNSURLRequest expects data not to be nil; see https://github.com/getsentry/sentry-cocoa/blob/394ad331f5d7c010ae68d884feda0a71dff5b194/Sources/Sentry/include/SentryNSURLRequest.h#L13-L24

I think this never was a problem because the SentryHttpTransport never tried to send envelopes that it couldn't serialize. So we definitely need a test to validate what happens in that edge case. Maybe we need to change the logic of SentryNSURLRequestBuilder to not create an `SentryNSURLRequest` when it can't serialize the envelope.

**Comment by denrase** [comment]
@philipphofmann Implemented handling of the case where envelope to data fails. Waiting for CI so we can do a final review/feedback round.

File: Sources/Sentry/SentryNSURLRequestBuilder.m
Comment (philipphofmann) [inline Sources/Sentry/SentryNSURLRequestBuilder.m:20]: `m`: As this is now essential functionality, please add a test for this. We don't have a `SentryNSURLRequestBuilder` test class yet, but I think we should add one.

**Review by philipphofmann — COMMENTED** [review]
We only miss tests for `SentryNSURLRequestBuilder`, then we can merge this. Thank you @denrase.

**Review by brustolin — APPROVED** [review]
Thanks @denrase, it looks good!

**Comment by philipphofmann** [comment]
Thank you @denrase and thanks @brustolin for approving the PR.

**Pull request merged**
